src/DynamicExpresso.Core/Parsing/Parser.cs

```diff
@@ -1207,41 +1207,134 @@ private Expression[] ParseArrayInitializerList()
+		private Expression ParseWithObjectInitializer(NewExpression newExpr, Type newType)
 		{
 			ValidateToken(TokenId.OpenCurlyBracket, ErrorMessages.OpenCurlyBracketExpected);
 			NextToken();
-			var bindings = ParseMemberInitializerList(newType);
+			var initializedInstance = ParseMemberAndInitializerList(newExpr, newType);
 			ValidateToken(TokenId.CloseCurlyBracket, ErrorMessages.CloseCurlyBracketExpected);
 			NextToken();
-			return bindings;
+			return initializedInstance;
 		}
 
-		private MemberBinding[] ParseMemberInitializerList(Type newType)
+		private Expression ParseMemberAndInitializerList(NewExpression newExpr, Type newType)
 		{
+			var originalPos = _token.pos;
 			var bindingList = new List<MemberBinding>();
+			var actions = new List<Expression>();
+			var instance = Expression.Variable(newType);
+			var allowCollectionInit = typeof(IEnumerable).IsAssignableFrom(newType);
 			while (true)
 			{
 				if (_token.id == TokenId.CloseCurlyBracket) break;
-				ValidateToken(TokenId.Identifier, ErrorMessages.IdentifierExpected);
+				if (_token.id != TokenId.Identifier)
+				{
+					ParseCollectionInitalizer(newType, originalPos, bindingList, actions, instance, allowCollectionInit);
+				}
+				else
+				{
+					ParsePossibleMemberBinding(newType, originalPos, bindingList, actions, instance, allowCollectionInit);
+				}
+				if (_token.id != TokenId.Comma) break;
+				NextToken();
+			}
+			if (bindingList.Count == 0)
+			{
+				actions.Insert(0, Expression.Assign(instance, newExpr));
+				actions.Add(instance);
+				return Expression.Block(new ParameterExpression[] { instance }, actions);
+			}
+			return Expression.MemberInit(newExpr, bindingList.ToArray());
+		}
 
-				var propertyOrFieldName = _token.text;
-				var member = FindPropertyOrField(newType, propertyOrFieldName, false);
-				if (member == null)
-					throw CreateParseException(_token.pos, ErrorMessages.UnknownPropertyOrField, propertyOrFieldName, GetTypeName(newType));
+		private void ParsePossibleMemberBinding(Type newType, int originalPos, List<MemberBinding> bindingList, List<Expression> actions, ParameterExpression instance, bool allowCollectionInit)
+		{
+			ValidateToken(TokenId.Identifier, ErrorMessages.IdentifierExpected);
 
+			var propertyOrFieldName = _token.text;
+			var member = FindPropertyOrField(newType, propertyOrFieldName, false);
+			var pos = _token.pos;
+			if (allowCollectionInit)
+			{
 				NextToken();
-
-				ValidateToken(TokenId.Equal, ErrorMessages.EqualExpected);
+				//new T(){Prop = 1}
+				//new T(){variable = 2}
+				if (_token.id == TokenId.Equal && member != null)
+				{
+					if (actions.Count > 0)
+					{
+						throw CreateParseException(pos, ErrorMessages.InvalidInitializerMemberDeclarator);
+					}
+				}
+				else if (_token.id != TokenId.Equal || _arguments.TryGetIdentifier(propertyOrFieldName, out _) || _arguments.TryGetParameters(propertyOrFieldName, out _))
+				{
+					SetTextPos(pos);
+					NextToken();
```

Comment:
Aren't these two operations NOP?

Reply:
If I take that out, it ends up causing the (now fixed) Ctor_NewListCantFindAddMethod test to fail with "Expression Expected" instead of cannot find appropriate Add method